jobs: TestRegistryLifecycle failed

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3250881&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3250881&tab=artifacts#/) on master @ [c995342ead51e08f8ed1155de4218d30a00d86d2](https://github.com/cockroachdb/cockroach/commits/c995342ead51e08f8ed1155de4218d30a00d86d2):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x516ba00]

```
Stack: 

```
goroutine 74085 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0021d6120, 0x8ea0ad0, 0xc0042069c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x5b5e540, 0xad5a700)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*Iterator).Valid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:107
github.com/cockroachdb/cockroach/pkg/sql.glob..func75(0x8ea0ad0, 0xc004206540, 0x8fd05e8, 0xc0042f4568, 0xc00109e598, 0x0, 0x0, 0xc0019bab00, 0xc0045d0960, 0xc0020d6aa0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1228 +0x1e0
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1(0x8e28660, 0xc00320f3e0, 0xc001902d38, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x5ae
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x8ea0ad0, 0xc0042069c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x1aa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc0021d6120, 0x8ea0ad0, 0xc0042069c0, 0xc0027f4000, 0x0, 0xc002319e30)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x162
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x28e
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegistryLifecycle/dump_traces_on_cancel
    jobs_test.go:250: Starting resume
    jobs_test.go:263: Exiting resume
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:299: Exiting OnFailOrCancel
E210802 06:30:21.191981 74082 jobs/adopt.go:327  [n1] 57  job 680969571288383489: adoption completed with error job canceled
E210802 06:30:21.200704 72865 jobs/registry.go:644  [-] 58  error claiming jobs: could not query jobs table: claim-jobs: context canceled
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-schema spaskob
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3250880&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3250880&tab=artifacts#/) on master @ [c995342ead51e08f8ed1155de4218d30a00d86d2](https://github.com/cockroachdb/cockroach/commits/c995342ead51e08f8ed1155de4218d30a00d86d2):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x34a6dad]

```
Stack: 

```
goroutine 68754 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00692ac00, 0x55dded0, 0xc009a95040)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x126
panic(0x3c94de0, 0x729c6a0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*Iterator).Valid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:107
github.com/cockroachdb/cockroach/pkg/sql.glob..func75(0x55dded0, 0xc009a94d40, 0x570c878, 0xc0068917c8, 0xc006637590, 0x0, 0x0, 0xc007b73d00, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1228 +0x12d
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1(0x5566780, 0xc00572db00, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x4a2
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x55dded0, 0xc009a95040)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x154
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc00692ac00, 0x55dded0, 0xc009a95040, 0xc004908180, 0x0, 0xc00950aa20)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegistryLifecycle/dump_traces_on_cancel
    jobs_test.go:250: Starting resume
    jobs_test.go:263: Exiting resume
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:299: Exiting OnFailOrCancel
E210802 06:31:26.817377 68724 jobs/adopt.go:327  [n1] 38  job 680969787040137217: adoption completed with error job canceled
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/sql-schema spaskob
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3254991&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3254991&tab=artifacts#/) on master @ [65d81716a22a38873a679b4fd5f1e019280725d5](https://github.com/cockroachdb/cockroach/commits/65d81716a22a38873a679b4fd5f1e019280725d5):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x34ab76d]

```
Stack: 

```
goroutine 68614 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc005ee4180, 0x55e4770, 0xc001da9b00)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x126
panic(0x3c9a120, 0x72a66a0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*Iterator).Valid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:107
github.com/cockroachdb/cockroach/pkg/sql.glob..func75(0x55e4770, 0xc001da9900, 0x5713118, 0xc005b3d4b8, 0xc002eeb590, 0x0, 0x0, 0xc005fe7180, 0x203001, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1228 +0x12d
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1(0x556cf60, 0xc003a74800, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x4a2
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x55e4770, 0xc001da9b00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x154
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc005ee4180, 0x55e4770, 0xc001da9b00, 0xc00235d440, 0x0, 0xc0042339b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegistryLifecycle/dump_traces_on_cancel
    jobs_test.go:250: Starting resume
    jobs_test.go:263: Exiting resume
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:299: Exiting OnFailOrCancel
E210803 06:38:02.376063 68589 jobs/adopt.go:327  [n1] 39  job 681254198670229505: adoption completed with error job canceled
E210803 06:38:02.378846 67538 jobs/registry.go:644  [-] 40  error claiming jobs: could not query jobs table: claim-jobs: context canceled
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/sql-schema spaskob
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3254992&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3254992&tab=artifacts#/) on master @ [65d81716a22a38873a679b4fd5f1e019280725d5](https://github.com/cockroachdb/cockroach/commits/65d81716a22a38873a679b4fd5f1e019280725d5):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5172c00]

```
Stack: 

```
goroutine 74716 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc002ee0ab0, 0x8eac710, 0xc002a57580)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x5b668a0, 0xad6a700)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*Iterator).Valid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:107
github.com/cockroachdb/cockroach/pkg/sql.glob..func75(0x8eac710, 0xc002a572c0, 0x8fdc228, 0xc0027988e8, 0xc002cf9598, 0x0, 0x0, 0xc0037c1180, 0x0, 0xc0033eeb40, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1228 +0x1e0
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1(0x8e341e0, 0xc00352a000, 0xc000660d38, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x5ae
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x8eac710, 0xc002a57580)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x1aa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc002ee0ab0, 0x8eac710, 0xc002a57580, 0xc0040313c0, 0x0, 0xc00275be60)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x162
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x28e
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegistryLifecycle/dump_traces_on_cancel
    jobs_test.go:250: Starting resume
    jobs_test.go:263: Exiting resume
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:299: Exiting OnFailOrCancel
E210803 06:43:23.584904 74593 jobs/adopt.go:327  [n1] 74  job 681255250468995073: adoption completed with error job canceled
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-schema spaskob
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3259506&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3259506&tab=artifacts#/) on master @ [eef03a46f2e43ff70485dadf7d9ad445db05cab4](https://github.com/cockroachdb/cockroach/commits/eef03a46f2e43ff70485dadf7d9ad445db05cab4):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x34b458d]

```
Stack: 

```
goroutine 68957 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00833de00, 0x5601090, 0xc00bf1d240)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x126
panic(0x3caaec0, 0x72cc6a0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*Iterator).Valid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:107
github.com/cockroachdb/cockroach/pkg/sql.glob..func75(0x5601090, 0xc00bf1d040, 0x572fbb8, 0xc00a2692f8, 0xc005dd7590, 0x0, 0x0, 0xc009a5bf80, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1228 +0x12d
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1(0x5589500, 0xc003a6b080, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x4a2
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x5601090, 0xc00bf1d240)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x154
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc00833de00, 0x5601090, 0xc00bf1d240, 0xc0033ec6c0, 0x0, 0xc00ab4b830)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegistryLifecycle/dump_traces_on_cancel
    jobs_test.go:250: Starting resume
    jobs_test.go:263: Exiting resume
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:299: Exiting OnFailOrCancel
E210804 06:33:34.993109 68972 jobs/adopt.go:327  [n1] 50  job 681536438033186817: adoption completed with error job canceled
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/sql-schema spaskob
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -2984,7 +2984,7 @@ func TestBackupRestoreDuringUserDefinedTypeChange(t *testing.T) {
 				ServerArgs: base.TestServerArgs{
 					Knobs: base.TestingKnobs{
 						SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-							RunBeforeEnumMemberPromotion: func() error {
+							RunBeforeEnumMemberPromotion: func(context.Context) error {
 								mu.Lock()
 								if numTypeChangesStarted < len(tc.queries) {
 									numTypeChangesStarted++

--- a/pkg/ccl/changefeedccl/BUILD.bazel
+++ b/pkg/ccl/changefeedccl/BUILD.bazel
@@ -168,6 +168,7 @@ go_test(
         "//pkg/server/status",
         "//pkg/server/telemetry",
         "//pkg/settings/cluster",
+        "//pkg/spanconfig",
         "//pkg/sql",
         "//pkg/sql/catalog",
         "//pkg/sql/catalog/catalogkv",

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -3177,6 +3177,7 @@ func TestChangefeedProtectedTimestamps(t *testing.T) {
 				select {
 				case <-waitUntilClosed:
 				case <-done:
+				case <-ctx.Done():
 				}
 			default:
 			}
@@ -3286,11 +3287,10 @@ func TestChangefeedProtectedTimestamps(t *testing.T) {
 				// canceled.
 				waitForBlocked = requestBlockedScan()
 				sqlDB.Exec(t, `ALTER TABLE foo ADD COLUMN d INT NOT NULL DEFAULT 2`)
-				unblock := waitForBlocked()
+				_ = waitForBlocked()
 				waitForRecord()
 				sqlDB.Exec(t, `CANCEL JOB $1`, foo.(cdctest.EnterpriseTestFeed).JobID())
 				waitForNoRecord()
-				unblock()
 			}
 		}, feedTestNoTenants, withArgsFn(func(args *base.TestServerArgs) {
 			storeKnobs := &kvserver.StoreTestingKnobs{}

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -34,6 +34,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
@@ -326,6 +327,7 @@ func startTestFullServer(
 	knobs := base.TestingKnobs{
 		DistSQL:          &execinfra.TestingKnobs{Changefeed: &TestingKnobs{}},
 		JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+		SpanConfig:       &spanconfig.TestingKnobs{ManagerDisableJobCreation: true},
 	}
 	if options.knobsFn != nil {
 		options.knobsFn(&knobs)

--- a/pkg/ccl/multiregionccl/region_test.go
+++ b/pkg/ccl/multiregionccl/region_test.go
@@ -9,6 +9,7 @@
 package multiregionccl_test
 
 import (
+	"context"
 	"sort"
 	"strings"
 	"testing"
@@ -119,7 +120,7 @@ func TestConcurrentAddDropRegions(t *testing.T) {
 
 			knobs := base.TestingKnobs{
 				SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-					RunBeforeEnumMemberPromotion: func() error {
+					RunBeforeEnumMemberPromotion: func(context.Context) error {
 						mu.Lock()
 						if firstOp {
 							firstOp = false
@@ -258,7 +259,7 @@ func TestRegionAddDropEnclosingRegionalByRowOps(t *testing.T) {
 
 				knobs := base.TestingKnobs{
 					SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-						RunBeforeEnumMemberPromotion: func() error {
+						RunBeforeEnumMemberPromotion: func(ctx context.Context) error {
 							mu.Lock()
 							defer mu.Unlock()
 							close(typeChangeStarted)
@@ -383,7 +384,7 @@ ALTER TABLE db.public.global CONFIGURE ZONE USING
 
 			knobs := base.TestingKnobs{
 				SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-					RunBeforeEnumMemberPromotion: func() error {
+					RunBeforeEnumMemberPromotion: func(context.Context) error {
 						close(regionOpStarted)
 						<-placementOpFinished
 						return nil
@@ -444,7 +445,7 @@ func TestSettingPrimaryRegionAmidstDrop(t *testing.T) {
 	dropRegionFinished := make(chan struct{})
 	knobs := base.TestingKnobs{
 		SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-			RunBeforeEnumMemberPromotion: func() error {
+			RunBeforeEnumMemberPromotion: func(context.Context) error {
 				mu.Lock()
 				defer mu.Unlock()
 				if dropRegionStarted != nil {
@@ -845,7 +846,7 @@ func TestRegionAddDropWithConcurrentBackupOps(t *testing.T) {
 
 				backupKnobs := base.TestingKnobs{
 					SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-						RunBeforeEnumMemberPromotion: func() error {
+						RunBeforeEnumMemberPromotion: func(ctx context.Context) error {
 							mu.Lock()
 							defer mu.Unlock()
 							if waitInTypeSchemaChangerDuringBackup {
@@ -902,7 +903,7 @@ INSERT INTO db.rbr VALUES (1,1),(2,2),(3,3);
 
 				restoreKnobs := base.TestingKnobs{
 					SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-						RunBeforeEnumMemberPromotion: func() error {
+						RunBeforeEnumMemberPromotion: func(context.Context) error {
 							mu.Lock()
 							defer mu.Unlock()
 							if !regionAlterCmd.shouldSucceed {

--- a/pkg/jobs/adopt.go
+++ b/pkg/jobs/adopt.go
@@ -325,33 +325,43 @@ func (r *Registry) resumeJob(ctx context.Context, jobID jobspb.JobID, s sqlliven
 	}
 	resumeCtx, cancel := r.makeCtx()
 
-	aj := &adoptedJob{sid: s.ID(), cancel: cancel}
-	r.addAdoptedJob(jobID, aj)
+	if alreadyAdopted := r.addAdoptedJob(jobID, s.ID(), cancel); alreadyAdopted {
+		return nil
+	}
+
 	r.metrics.ResumedJobs.Inc(1)
 	if err := r.stopper.RunAsyncTask(ctx, job.taskName(), func(ctx context.Context) {
 		// Wait for the job to finish. No need to print the error because if there
 		// was one it's been set in the job status already.
 		_ = r.runJob(resumeCtx, resumer, job, status, job.taskName())
 	}); err != nil {
-		r.removeAdoptedJob(jobID)
+		r.unregister(jobID)
 		return err
 	}
 	return nil
 }
 
-func (r *Registry) removeAdoptedJob(jobID jobspb.JobID) {
+// addAdoptedJob adds the job to the set of currently running jobs. This set is
+// used for introspection, and, importantly, to serve as a lock to prevent
+// concurrent executions. Removal occurs in runJob or in the case that we were
+// unable to launch the goroutine to call runJob. If the returned boolean is
+// false, it means that the job is already registered as running and should not
+// be run again.
+func (r *Registry) addAdoptedJob(
+	jobID jobspb.JobID, sessionID sqlliveness.SessionID, cancel context.CancelFunc,
+) (alreadyAdopted bool) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	delete(r.mu.adoptedJobs, jobID)
-}
 
-func (r *Registry) addAdoptedJob(jobID jobspb.JobID, aj *adoptedJob) {
-	// TODO(sajjad): We should check whether adoptedJobs already has jobID or not. If
-	// the ID exists, we should not add it again and the caller should not start
-	// another resumer.
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	r.mu.adoptedJobs[jobID] = aj
+	if _, alreadyAdopted = r.mu.adoptedJobs[jobID]; alreadyAdopted {
+		return true
+	}
+
+	r.mu.adoptedJobs[jobID] = &adoptedJob{
+		sid:    sessionID,
+		cancel: cancel,
+	}
+	return false
 }
 
 func (r *Registry) runJob(
@@ -455,11 +465,11 @@ func (r *Registry) servePauseAndCancelRequests(ctx context.Context, s sqllivenes
 			statusString := *row[1].(*tree.DString)
 			switch Status(statusString) {
 			case StatusPaused:
-				r.unregister(id)
+				r.cancelRegisteredJobContext(id)
 				log.Infof(ctx, "job %d, session %s: paused", id, s.ID())
 			case StatusReverting:
 				if err := job.Update(ctx, txn, func(txn *kv.Txn, md JobMetadata, ju *JobUpdater) error {
-					r.unregister(id)
+					r.cancelRegisteredJobContext(id)
 					md.Payload.Error = errJobCanceled.Error()
 					encodedErr := errors.EncodeError(ctx, errJobCanceled)
 					md.Payload.FinalResumeError = &encodedErr

--- a/pkg/jobs/jobs.go
+++ b/pkg/jobs/jobs.go
@@ -862,7 +862,7 @@ func (j *Job) getRunStats() (rs RunStats) {
 // and nothing will be send on errCh. Clients must not start jobs more than
 // once.
 func (sj *StartableJob) Start(ctx context.Context) (err error) {
-	if starts := atomic.AddInt64(&sj.starts, 1); starts != 1 {
+	if alreadyStarted := sj.recordStart(); alreadyStarted {
 		return errors.AssertionFailedf(
 			"StartableJob %d cannot be started more than once", sj.ID())
 	}
@@ -960,8 +960,19 @@ func (sj *StartableJob) CleanupOnRollback(ctx context.Context) error {
 // Cancel will mark the job as canceled and release its resources in the
 // Registry.
 func (sj *StartableJob) Cancel(ctx context.Context) error {
-	defer sj.registry.unregister(sj.ID())
+	alreadyStarted := sj.recordStart() // prevent future start attempts
+	defer func() {
+		if alreadyStarted {
+			sj.registry.cancelRegisteredJobContext(sj.ID())
+		} else {
+			sj.registry.unregister(sj.ID())
+		}
+	}()
 	return sj.registry.CancelRequested(ctx, nil, sj.ID())
+}
+
+func (sj *StartableJob) recordStart() (alreadyStarted bool) {
+	return atomic.AddInt64(&sj.starts, 1) != 1
 }
 
 // FormatRetriableExecutionErrorLogToStringArray extracts the events

--- a/pkg/jobs/jobs_test.go
+++ b/pkg/jobs/jobs_test.go
@@ -35,6 +35,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
@@ -44,7 +45,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -222,6 +222,9 @@ func (rts *registryTestSuite) setUp(t *testing.T) {
 			}
 		}
 		args.Knobs.JobsTestingKnobs = knobs
+		args.Knobs.SpanConfig = &spanconfig.TestingKnobs{
+			ManagerDisableJobCreation: true,
+		}
 
 		if rts.traceRealSpan {
 			baseDir, dirCleanupFn := testutils.TempDir(t)
@@ -283,6 +286,11 @@ func (rts *registryTestSuite) setUp(t *testing.T) {
 			},
 			FailOrCancel: func(ctx context.Context) error {
 				t.Log("Starting FailOrCancel")
+				if rts.traceRealSpan {
+					// Add a dummy recording so we actually see something in the trace.
+					span := tracing.SpanFromContext(ctx)
+					span.RecordStructured(&types.StringValue{Value: "boom"})
+				}
 				rts.mu.Lock()
 				rts.mu.a.OnFailOrCancelStart = true
 				rts.mu.Unlock()
@@ -362,7 +370,6 @@ func (rts *registryTestSuite) check(t *testing.T, expectedStatus jobs.Status) {
 
 func TestRegistryLifecycle(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 68315, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	t.Run("normal success", func(t *testing.T) {
@@ -1095,42 +1102,40 @@ func TestRegistryLifecycle(t *testing.T) {
 	})
 
 	t.Run("dump traces on cancel", func(t *testing.T) {
-		rts := registryTestSuite{traceRealSpan: true}
+		completeCh := make(chan struct{})
+		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func() {
+			completeCh <- struct{}{}
+		}}
 		rts.setUp(t)
 		defer rts.tearDown()
-
-		runJobAndFail := func(expectedNumFiles int) {
-			j, err := jobs.TestingCreateAndStartJob(context.Background(), rts.registry, rts.s.DB(), rts.mockJob)
-			if err != nil {
-				t.Fatal(err)
-			}
-			rts.job = j
-
-			rts.mu.e.ResumeStart = true
-			rts.resumeCheckCh <- struct{}{}
-			rts.check(t, jobs.StatusRunning)
-
-			rts.sqlDB.Exec(t, "CANCEL JOB $1", j.ID())
-
-			// Cancellation will cause the running instance of the job to get context
-			// canceled causing it to potentially dump traces.
-			require.Error(t, rts.job.AwaitCompletion(rts.ctx))
-			checkTraceFiles(t, rts.registry, expectedNumFiles)
-
-			rts.mu.e.OnFailOrCancelStart = true
-			rts.check(t, jobs.StatusReverting)
-
-			rts.failOrCancelCheckCh <- struct{}{}
-			close(rts.failOrCancelCheckCh)
-			rts.failOrCancelCh <- nil
-			close(rts.failOrCancelCh)
-			rts.mu.e.OnFailOrCancelExit = true
-
-			rts.check(t, jobs.StatusCanceled)
-		}
-
 		rts.sqlDB.Exec(t, `SET CLUSTER SETTING jobs.trace.force_dump_mode='onStop'`)
-		runJobAndFail(1)
+		j, err := jobs.TestingCreateAndStartJob(rts.ctx, rts.registry, rts.s.DB(), rts.mockJob)
+		if err != nil {
+			t.Fatal(err)
+		}
+		rts.job = j
+
+		rts.mu.e.ResumeStart = true
+		rts.resumeCheckCh <- struct{}{}
+		rts.check(t, jobs.StatusRunning)
+
+		rts.sqlDB.Exec(t, "CANCEL JOB $1", j.ID())
+
+		<-completeCh
+		checkTraceFiles(t, rts.registry, 1)
+
+		rts.mu.e.OnFailOrCancelStart = true
+		rts.check(t, jobs.StatusReverting)
+
+		rts.failOrCancelCheckCh <- struct{}{}
+		close(rts.failOrCancelCheckCh)
+		rts.failOrCancelCh <- nil
+		close(rts.failOrCancelCh)
+		rts.mu.e.OnFailOrCancelExit = true
+
+		rts.check(t, jobs.StatusCanceled)
+
+		<-completeCh
 	})
 }
 
@@ -1754,7 +1759,7 @@ func TestShowJobs(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	params, _ := tests.CreateTestServerParams()
-	params.Knobs = base.TestingKnobs{JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals()}
+	params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
 	s, rawSQLDB, _ := serverutils.StartServer(t, params)
 	sqlDB := sqlutils.MakeSQLRunner(rawSQLDB)
 	ctx := context.Background()
@@ -2583,11 +2588,9 @@ func TestStartableJob(t *testing.T) {
 		status, err := sj.CurrentStatus(ctx, nil /* txn */)
 		require.NoError(t, err)
 		require.Equal(t, jobs.StatusCancelRequested, status)
-		for _, id := range jr.CurrentlyRunningJobs() {
-			require.NotEqual(t, id, sj.ID())
-		}
+		// Start should fail since we have already called cancel on the job.
 		err = sj.Start(ctx)
-		require.Regexp(t, "job with status cancel-requested cannot be marked started", err)
+		require.Regexp(t, "cannot be started more than once", err)
 	})
 	setUpRunTest := func(t *testing.T) (
 		sj *jobs.StartableJob,

--- a/pkg/jobs/registry.go
+++ b/pkg/jobs/registry.go
@@ -652,12 +652,13 @@ func (r *Registry) CreateStartableJobWithTxn(
 		// Using a new context allows for independent lifetimes and cancellation.
 		resumerCtx, cancel = r.makeCtx()
 
-		r.mu.Lock()
-		defer r.mu.Unlock()
-		if _, alreadyRegistered := r.mu.adoptedJobs[jobID]; alreadyRegistered {
-			log.Fatalf(ctx, "job %d: was just created but found in registered adopted jobs", jobID)
+		if alreadyAdopted := r.addAdoptedJob(jobID, j.sessionID, cancel); alreadyAdopted {
+			log.Fatalf(
+				ctx,
+				"job %d: was just created but found in registered adopted jobs",
+				jobID,
+			)
 		}
-		r.mu.adoptedJobs[jobID] = &adoptedJob{sid: j.sessionID, cancel: cancel}
 		execDone = make(chan struct{})
 	}
 
@@ -1340,14 +1341,17 @@ func (r *Registry) cancelAllAdoptedJobs() {
 func (r *Registry) unregister(jobID jobspb.JobID) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-
-	aj, ok := r.mu.adoptedJobs[jobID]
-	// It is possible for a job to be double unregistered. unregister is always
-	// called at the end of resume. But it can also be called during deprecatedCancelAll
-	// and in the adopt loop under certain circumstances.
-	if ok {
+	if aj, ok := r.mu.adoptedJobs[jobID]; ok {
 		aj.cancel()
 		delete(r.mu.adoptedJobs, jobID)
+	}
+}
+
+func (r *Registry) cancelRegisteredJobContext(jobID jobspb.JobID) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if aj, ok := r.mu.adoptedJobs[jobID]; ok {
+		aj.cancel()
 	}
 }
 

--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -72,6 +72,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/collector"
 	"github.com/cockroachdb/errors"
 )
 
@@ -1283,7 +1284,8 @@ CREATE TABLE crdb_internal.cluster_inflight_traces (
 		}
 
 		traceCollector := p.ExecCfg().TraceCollector
-		for iter := traceCollector.StartIter(ctx, traceID); iter.Valid(); iter.Next() {
+		var iter *collector.Iterator
+		for iter, err = traceCollector.StartIter(ctx, traceID); err == nil && iter.Valid(); iter.Next() {
 			nodeID, recording := iter.Value()
 			traceString := recording.String()
 			traceJaegerJSON, err := recording.ToJaegerJSON("", "", fmt.Sprintf("node %d", nodeID))
@@ -1297,6 +1299,12 @@ CREATE TABLE crdb_internal.cluster_inflight_traces (
 				tree.NewDString(traceJaegerJSON)); err != nil {
 				return false, err
 			}
+		}
+		if err != nil {
+			return false, err
+		}
+		if iter.Error() != nil {
+			return false, iter.Error()
 		}
 
 		return true, nil

--- a/pkg/sql/type_change.go
+++ b/pkg/sql/type_change.go
@@ -180,7 +180,7 @@ type TypeSchemaChangerTestingKnobs struct {
 	RunBeforeExec func() error
 	// RunBeforeEnumMemberPromotion runs before enum members are promoted from
 	// readable to all permissions in the typeSchemaChanger.
-	RunBeforeEnumMemberPromotion func() error
+	RunBeforeEnumMemberPromotion func(ctx context.Context) error
 	// RunAfterOnFailOrCancel runs after OnFailOrCancel completes, if
 	// OnFailOrCancel is triggered.
 	RunAfterOnFailOrCancel func() error
@@ -272,7 +272,7 @@ func (t *typeSchemaChanger) exec(ctx context.Context) error {
 		typeDesc.GetKind() == descpb.TypeDescriptor_MULTIREGION_ENUM) &&
 		len(t.transitioningMembers) != 0 {
 		if fn := t.execCfg.TypeSchemaChangerTestingKnobs.RunBeforeEnumMemberPromotion; fn != nil {
-			if err := fn(); err != nil {
+			if err := fn(ctx); err != nil {
 				return err
 			}
 		}

--- a/pkg/sql/type_change_test.go
+++ b/pkg/sql/type_change_test.go
@@ -13,7 +13,6 @@ package sql_test
 import (
 	"context"
 	"fmt"
-	"sync"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/jobs"
@@ -517,18 +516,18 @@ func TestTypeChangeJobCancelSemantics(t *testing.T) {
 			params, _ := tests.CreateTestServerParams()
 
 			// Wait groups for synchronizing various parts of the test.
-			var typeSchemaChangeStarted sync.WaitGroup
-			typeSchemaChangeStarted.Add(1)
-			var blockTypeSchemaChange sync.WaitGroup
-			blockTypeSchemaChange.Add(1)
-			var finishedSchemaChange sync.WaitGroup
-			finishedSchemaChange.Add(1)
+			typeSchemaChangeStarted := make(chan struct{})
+			blockTypeSchemaChange := make(chan struct{})
+			finishedSchemaChange := make(chan struct{})
 
 			params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
 			params.Knobs.SQLTypeSchemaChanger = &sql.TypeSchemaChangerTestingKnobs{
-				RunBeforeEnumMemberPromotion: func() error {
-					typeSchemaChangeStarted.Done()
-					blockTypeSchemaChange.Wait()
+				RunBeforeEnumMemberPromotion: func(ctx context.Context) error {
+					close(typeSchemaChangeStarted)
+					select {
+					case <-blockTypeSchemaChange:
+					case <-ctx.Done():
+					}
 					return nil
 				},
 			}
@@ -555,10 +554,10 @@ CREATE TYPE db.greetings AS ENUM ('hi', 'yo');
 				if !tc.cancelable && err != nil {
 					t.Error(err)
 				}
-				finishedSchemaChange.Done()
+				close(finishedSchemaChange)
 			}()
 
-			typeSchemaChangeStarted.Wait()
+			<-typeSchemaChangeStarted
 
 			_, err = sqlDB.Exec(`CANCEL JOB (
 SELECT job_id FROM [SHOW JOBS]
@@ -594,8 +593,8 @@ WHERE
 					return nil
 				})
 			}
-			blockTypeSchemaChange.Done()
-			finishedSchemaChange.Wait()
+			close(blockTypeSchemaChange)
+			<-finishedSchemaChange
 		})
 	}
 }

--- a/pkg/util/tracing/collector/collector.go
+++ b/pkg/util/tracing/collector/collector.go
@@ -87,11 +87,12 @@ type Iterator struct {
 
 // StartIter fetches the live nodes in the cluster, and configures the underlying
 // Iterator that is used to access recorded spans in a streaming fashion.
-func (t *TraceCollector) StartIter(ctx context.Context, traceID uint64) *Iterator {
+func (t *TraceCollector) StartIter(ctx context.Context, traceID uint64) (*Iterator, error) {
 	tc := &Iterator{ctx: ctx, traceID: traceID, collector: t}
-	tc.liveNodes, tc.iterErr = nodesFromNodeLiveness(ctx, t.nodeliveness)
-	if tc.iterErr != nil {
-		return nil
+	var err error
+	tc.liveNodes, err = nodesFromNodeLiveness(ctx, t.nodeliveness)
+	if err != nil {
+		return nil, err
 	}
 
 	// Calling Next() positions the Iterator in a valid state. It will fetch the
@@ -99,7 +100,7 @@ func (t *TraceCollector) StartIter(ctx context.Context, traceID uint64) *Iterato
 	// nodes.
 	tc.Next()
 
-	return tc
+	return tc, nil
 }
 
 // Valid returns whether the Iterator is in a valid state to read values from.

--- a/pkg/util/tracing/collector/collector_test.go
+++ b/pkg/util/tracing/collector/collector_test.go
@@ -133,10 +133,12 @@ func TestTracingCollectorGetSpanRecordings(t *testing.T) {
 		res := make(map[roachpb.NodeID][]tracing.Recording)
 
 		var iter *collector.Iterator
-		for iter = traceCollector.StartIter(ctx, traceID); iter.Valid(); iter.Next() {
+		var err error
+		for iter, err = traceCollector.StartIter(ctx, traceID); err == nil && iter.Valid(); iter.Next() {
 			nodeID, recording := iter.Value()
 			res[nodeID] = append(res[nodeID], recording)
 		}
+		require.NoError(t, err)
 		require.NoError(t, iter.Error())
 		return res
 	}

